### PR TITLE
BE Pay Route

### PR DESCRIPTION
**Pull request by jb162**
### What this PR does (required):
- addresses #32
- charges client and transfers payment to sitter 
   - clientId, clientPaymentMethod, chargeAmount (from client), payoutAmount (to sitter), and transferGroup are passed in req.body
      - note: amounts in Stripe are in cents, (e.g. 2800 = $28)
   - sitterId is passed in req.params as id

### Any information needed to test this feature (required):
- create .env with STRIPE_API_KEY variable (created new account, new key in Slack)
- run server in terminal
```
npm run dev
```
- sent POST request to http://localhost:3001/stripe/request/cus_KrskO3XeQcNt2e/pay via Postman
   - if using your own Stripe account, change clientId, clientPaymentMethod, and URL id parameter using two customers registered on your own Stripe account
image

### Any issues with the current functionality (optional):
- currently, the client is able to be charged but the payment is not transferred to the sitter
Screen Shot 2021-12-29 at 1 19 34 PM

- screenshot of client customer in Stripe

image

- screenshot of sitter customer in Stripe

image

### CODE REVIEW

File: server/controllers/stripe.js [hunk 1]
```diff
@@ -45,29 +45,51 @@ exports.retrieveCustomer = async (req, res) => {
...
-    type: "card",
-    card: {
-      number: "4242424242424242",
-      exp_month: 12,
-      exp_year: 2022,
-      cvc: "314",
-    },
-  });
-  res.send(paymentMethod);
+exports.retrieveAllCustomers = async (req, res) => {
```
Comment (rajivtitus) [inline server/controllers/stripe.js:51]: I don't think we'll need a route to retrieve all of the customers? :)

File: server/controllers/stripe.js [hunk 2]
```diff
@@ -45,29 +45,51 @@ exports.retrieveCustomer = async (req, res) => {
...
 };
 
-// @route GET /stripe/payment/all
+// @route GET /stripe/payment/all/:customerId
 // @desc gets all payment methods
 // @access Public
 exports.getAllPaymentMethods = async (req, res) => {
   const paymentMethods = await stripe.paymentMethods.list({
-    customer: "cus_Kk1rd4KHOzlRC0",
+    customer: req.params.customerId,
```
Comment (rajivtitus) [inline server/controllers/stripe.js:61]: We should restrict the retrieval of payment methods only to the logged in user!

File: server/controllers/stripe.js (hunk 1 above)
Comment (jb162) [reply server/controllers/stripe.js:51]: I just added this in for testing with Postman I'll remove it when the PR is fully functional :)

File: server/controllers/stripe.js (hunk 2 above)
Comment (jb162) [reply server/controllers/stripe.js]: I fixed this by using req.user.stripeId to access the logged-in customer's payment method (had to add protect middleware to the "GET /stripe/payment/all/:customerId" route). 

To add stripeId to User, I had to modify the User model (to include stripeId parameter) and "POST /auth/register" route  (to create Stripe customer upon register)

File: server/routes/stripe.js [hunk 3]
```diff
@@ -2,16 +2,17 @@ const express = require("express");
...
 } = require("../controllers/stripe");
+const protect = require("../middleware/auth");
 
 router.route("/session").post(createSession);
-router.route("/customers/create").post(createCustomer);
 router.route("/customers/retrieve/:id").get(retrieveCustomer);
-router.route("/payment/create").post(createPaymentMethod);
-router.route("/payment/all").get(getAllPaymentMethods);
+router.route("/customers/retrieveAll").get(retrieveAllCustomers);
+router.route("/payment/all/:customerId").get(protect, getAllPaymentMethods);
```
Comment (rajivtitus) [inline server/routes/stripe.js:15]:
```suggestion
router.route("/payment/all").get(protect, getAllPaymentMethods);
```
We no longer require that path param 😄

File: server/controllers/stripe.js [hunk 4]
```diff
@@ -45,29 +35,51 @@ exports.retrieveCustomer = async (req, res) => {
...
 
-// @route GET /stripe/payment/all
+// @route GET /stripe/payment/all/:customerId
 // @desc gets all payment methods
-// @access Public
+// @access Private
 exports.getAllPaymentMethods = async (req, res) => {
   const paymentMethods = await stripe.paymentMethods.list({
-    customer: "cus_Kk1rd4KHOzlRC0",
+    customer: req.user.stripeId,
```
Comment (rajivtitus) [inline server/controllers/stripe.js:51]: When we perform the authentication in our middleware, we need to add the `stripeId` to the request object as well, I don't think we are currently doing that!

Please see: `/server/middleware/auth.js`

**Review by rajivtitus — COMMENTED** [review]
Just made a couple of comments based on the new changes :)

File: server/controllers/stripe.js (hunk 4 above)
Comment (jb162) [reply server/controllers/stripe.js]: I couldn't figure out how to change the auth middleware, as I am unfamiliar with jwt. Instead, I added in a call to find the user via req.user.id, and then used user.stripeId inside the stripe call

**Review by rajivtitus — APPROVED** [review]
LGTM 👍🏽

**Pull request merged**